src/uwtools/drivers/ww3.py

```diff
@@ -0,0 +1,97 @@
+"""
+A driver for ww3.
+"""
+
+from pathlib import Path
+from typing import List, Optional
+
+import f90nml  # type: ignore
+from iotaa import asset, task, tasks
+
+from uwtools.config.formats.nml import NMLConfig
+from uwtools.drivers.driver import Driver
+from uwtools.exceptions import UWConfigError
+from uwtools.strings import STR
+
+
+class WaveWatchIII(Driver):
+    """
+    A driver for ww3.
+    """
+
+    def __init__(
+        self,
+        config: Optional[Path] = None,
+        dry_run: bool = False,
+        batch: bool = False,
+        key_path: Optional[List[str]] = None,
+    ):
+        """
+        The driver.
+
+        :param config: Path to config file (read stdin if missing or None).
+        :param dry_run: Run in dry-run mode?
+        :param batch: Run component via the batch system?
+        :param key_path: Keys leading through the config to the driver's configuration block.
+        """
+        super().__init__(config=config, dry_run=dry_run, batch=batch, key_path=key_path)
+
+    # Workflow tasks
+
+    @task
+    def namelist_file(self):
+        """
+        The namelist file.
+        """
+        path = self._rundir / "namelists.nml"
+        yield self._taskname(str(path))
+        yield asset(path, path.is_file)
+        yield None
+        try:
+            namelist = self._driver_config["namelist"]
+        except KeyError as e:
+            raise UWConfigError(
+                "Provide either a 'namelist' YAML block or the %s file" % path
+            ) from e
+        additional_files = namelist.get("additional_files", {})
+        for file in additional_files:
```

Comment:
The logic here seems to have become convoluted. I'm torn between making this a 'softlink all specified namelists to the rundir', or fixing this so that it reads all the namelists and generates a merged one correctly?

Reply:
It would help me, at least, to distinguish between "namelists" and "namelist files": A namelist file can contain multiple namelists, and a namelist does't have to be in a file. And there may be different component- or app-imposed requirements around required names for namelists and files.

What are WW3's requirements w.r.t. namelists and namelist files? I mean, is it expecting to open one file, multiple files? Are the filenames hardcoded/predefined, or do we have control over filenames at run time? Are we certain there's no possibility of conflict between namelist names? Because, if not, concatenating namelists into a single file may not be safe.

I'm also looking at the JSON Schema for this component and trying to understand the use case. Could you describe that, or point to a design doc if I missed/forgot that? It seems that there are potentially multiple namelist files in play, and I don't see any support for `update_values` -- do we need that?

Reply:
Unfortunately, its not terribly friendly regarding the file components. There are [11 namelists](https://github.com/NOAA-EMC/WW3/tree/develop/model/nml) that are used in the various applications and configurations. There are [different](https://github.com/NOAA-EMC/WW3/blob/14e6bedfcefeff8f0896acea5dade4100a3fbacd/model/bin/README.md) default parameters for some operational configurations. Global Workflow walks through the namelists in a [script](https://github.com/NOAA-EMC/global-workflow/blob/50c2b8951b29a3c883a778becbf8582f9519eb48/ush/parsing_namelists_WW3.sh#L4) so I thought it would be important for us to know and track the namelists used

Reply:
I left `update_values` out as I was not familiar with any configurations where the user would override the values externally, and it makes it more complicated as well

Reply:
Those are 11 files files, but more than 11 namelists -- some of the files contain multiple namelists.

I see that some of the namelists (e.g. [this one](https://github.com/NOAA-EMC/WW3/blob/develop/model/nml/ww3_multi.nml#L145-L179)) contain both array-element _and_ object member syntax, and AFAIK there's no way to directly express that in YAML, though we could with some effort define a representation in YAML that could then map on to namelist syntax with custom logic. We haven't done this yet, though, as we've only dealt with simple namelist values.

So, this looks complicated. Would it be reasonable to require users to provide ready-to-go namelist files and have a task called something like `namelist_files()` that behaved similarly to, say, `files_copied()` from the `fv3` driver? If those 11 files with those specific names are required, we could enforce that via the schema.

Reply:
Fair! WWIII also uses the file-namelist and namelist-namelist interchangeably, which is...ok, really more reasons to strangle away from Fortran where possible. That might go back to my earlier thought of just listing the components needed and marking the necessary files. In Global, it looks like
```# Build exes for prep jobs and post jobs:
prep_exes="ww3_grid ww3_prep ww3_prnc ww3_grid"
post_exes="ww3_outp ww3_outf ww3_outp ww3_gint ww3_ounf ww3_ounp ww3_grib"
```
The `YAML` might then be:
```ww3: 
  config: 
    switch: 
    prep_exes: 
    post_exes: 
    run_dir: 
```
And just use `files_copied()` here as well

Reply:
What would be the benefit of the `config:` level there? Isn't this all "config"? Or is there something _besides_ `config:` at that level that would need to be kept separate?

Reply:
Ah, force of habit? I admit an attachment to that subblock, but it is sort of unnecessary here

Reply:
If we look at the couple of options that are necessary in "library" mode taking the ufs-weather-model regression test suite as an example, I'm seeing only one option for running WW3 in coupled mode, and that's the ww3_shel.nml.

After diving deeper into the fortran code that's used in the ufs-weather-model, there are two options for providing this configuration. An old non-standard configuration file that's managed with a template: https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/parm/ww3_shel.inp.IN

The same code also promises to use a standard nml file, if it's present with the name `ww3_shel.nml`.

I'd say we support the namelist option and try to find out more about why the ufs-weather-model doesn't use it.

The code that does the work of parsing these two files is here: https://github.com/NOAA-EMC/WW3/blob/d9b3172f4197c65d471662c6952a668152d71230/model/src/wav_shel_inp.F90

We should not worry about any of the 11 other namelist files and focus only on this one.

Reply:
@maddenp-noaa If we stick to the one namelist file, I feel like we're going to have some of the same complications that you mentioned surrounding array element and object member syntax. Can we take a short-cut for now to go the flat template route until we can come back to this more complicated problem in the future?

Reply:
Thanks @christinaholtNOAA , its so easy to get into rabbit holes with this one

Reply:
@christinaholtNOAA I think templating is the only way to handle these data types today, though we should probably think about a generic way to handle these in the future -- a UW YAML syntax for these kinds of LHS names -- as I don't think we can avoid this forever.

(BTW I think I was wrong in calling the ones we see here "object member syntax". There _is_ object-oriented Fortran, but I think this is just an example of a [user-defined type](https://www.cs.emory.edu/~cheung/Courses/561/Syllabus/6-Fortran/struct.html), which is much more common, in my experience, in NWP Fortran than is actual OO constructs.)

The `f90nml` representation might provide a guide, e.g.
```
>>> f90nml.reads("&nl o(2)%m(4:5)=88,99 /").todict()
OrderedDict({'nl': OrderedDict({'o': [OrderedDict({'m': [88, 99], '_start_index': {'m': [4]}})], '_start_index': {'o': [2]}})})
```
The namelist syntax says that there's an array `o` of some user-defined type that has a member called `m`, which is another array; and we're setting the 4th and 5th elements of `m` on the 2nd element of `o` to the integers `88` and `99`. And that's how `f90nml` keeps track of all that internally. So we could in theory design a UW YAML representation of that, then detect and turn into `f90nml` `Namelist` data, then let `f90nml` serialize back to namelist-formatted data. We might flag such values with a custom YAML tag so that they'd get the proper treatment.

Reply:
Would it be reasonable to cut this down to the UFS namelist for now, then spin off the syntax as another ticket? I think theres enough time in the sprint to knock that out while its on our minds too.

Reply:
@WeirAE I was just musing there about how we could eventually support those Fortran namelist constructs in YAML, not suggesting that we implement it now.

Reply:
Totally agree with @maddenp-noaa. This is a nice to have that we can include in our backlog, but let's just go the templating route for this work.